### PR TITLE
[CALCITE-5362] Geometry measurement functions

### PR DESCRIPTION
**Pull request by bchapuis**
This PR adds support for the following spatial measurement functions:
- ST_Area
- ST_ClosestCoordinate
- ST_ClosestPoint
- ST_FurthestCoordinate
- ST_Length
- ST_LocateAlong
- ST_LongestLine
- ST_MaxDistance
- ST_Perimeter
- ST_ProjectPoint

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java [hunk 1]
```diff
@@ -1318,6 +1322,195 @@ public static Geometry ST_Translate(Geometry geom, BigDecimal x, BigDecimal y) {
     return transformation.transform(geom);
   }
 
+  // Geometry measurement functions
+
+  /**
+   * Returns the area of the {@code geom}.
+   */
+  public static @Nullable Double ST_Area(Geometry geom) {
+    return geom.getArea();
```
Comment (rubenada) [inline core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java:1331]: Just curious: why `ST_Area` and  `ST_Length` don't have a null-check and the other methods do?

File: core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java [hunk 2]
```diff
@@ -1318,6 +1322,195 @@ public static Geometry ST_Translate(Geometry geom, BigDecimal x, BigDecimal y) {
...
+    for (int i = 0; i < geom.getNumGeometries(); i++) {
+      Geometry geometry = geom.getGeometryN(i);
+      if (geometry instanceof Polygon) {
+        perimeter += geometry.getLength();
+      }
+    }
+    return perimeter;
+  }
+
+  public static @Nullable Geometry ST_ProjectPoint(Geometry point, Geometry line) {
```
Comment (rubenada) [inline core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java:1501]: Minor: javadoc seems to be missing here.

File: core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java (hunk 1 above)
Comment (bchapuis) [reply core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java:1331]: This is a good point, I fixed these two methods. I started adding test cases for null values after realizing that they were supported by postgis. I will open an issue to ensure that we are consistent when it come to null values in the other methods.

File: core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java (hunk 2 above)
Comment (bchapuis) [reply core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java]: Thank you, I added the javadoc.

File: core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java (hunk 1 above)
Comment (bchapuis) [reply core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java:1331]: After reading a the javadoc for [SqlFunctions](https://javadoc.io/doc/org.apache.calcite/calcite-core/1.26.0/org/apache/calcite/runtime/SqlFunctions.html), it looks like checks for nulls are performed by Calcite before calling the functions. From what I understand, the default behavior is as follow: the functions always return null if one or more of its arguments are null. As this behavior looks similar to what postgis does, I removed all my null checks. Can you confirm that this is the right approach?

Comment (bchapuis) [reply core/src/main/java/org/apache/calcite/runtime/SpatialTypeFunctions.java:1331]: Thanks a lot for the clarification and the pointers.

**Pull request merged**